### ISSUE TITLE
[Request] Support for weird SS2 audio files from Drakengard 1

### ISSUE THREAD

**Issue by Surihix** [vgmstream/vgmstream#1238]
I decided to dump the music files of Drakengard 1 and I was able to get each individual music tracks from the game's archive files extracted neatly. 

I saw the header starting with SShd for each audio track and I assumed it is a standard .ss2 audio file which is common in a lot of PS2 games (from what I have noticed). so I saved the files in that extension.

But as I try converting and playing these ss2 tracks with vgmstream, I hear the audio sounding like a complete mess. 
I initially tested with foobar2000 with vgmstream component and heard the audio sounding quite badly. then I tried converting the track with vgmstream's test exe and the converted audio file also resulted in the same messy sounding audio.

I decided to use another tool called PSound (which you can download from the link below) for checking my extracted audio track as I was worried if my extraction process of the audio track from the game's files itself was incorrect.
http://snailrush.online.fr/

With PSound, the audio plays correctly and even converts properly too. so my audio extraction process from the game's archive files was correct and the problem was from vgmstream. 

Here is the ss2 audio file:
[track1.zip](https://github.com/vgmstream/vgmstream/files/9690937/track1.zip)

And here is the track that is converted with PSound to wav. you can use this for checking purposes:
https://drive.google.com/file/d/1-tPASXgQS9czHxb4jAj3K3MW1iyCy8Je/view?usp=sharing

**Comment by Hexagon123** [comment]
This seems somewhat badly (poorly) extracted?

**Comment by Surihix** [comment]
> This seems somewhat badly (poorly) extracted?

Yeah that's what I thought at first as well but PSound is able to play and convert these tracks just fine.

**Comment by bnnm** [comment]
vgmstream accepts known "real" formats (header + data) and follows info in the header, while psound simply reads raw data (ignoring headers) and does heuristics to autodetect PS-ADPCM data. So, psound playing something isn't a good comparison for us.

The file you posted isn't correctly ripped, as it starts with the "SShd" header, and ends with "cavia stream format v1.01" + blank data. Proper cavia sound files should start with "cavia stream format v1.01" + blank, and then "SShd". Example: [SNDIMGEU_001.zip](https://github.com/vgmstream/vgmstream/files/9692974/SNDIMGEU_001.zip)

vgmstream just follows header values, so unless "cavia stream.." goes first data is misaligned (start is set to 0x800, from "cavia stream"). Files also need to be named .ads (SShd's actual extension since .ss2 is fake) but I guess I could allow .ss2 too.

(there is a table in the .BIN bigfile that gives you proper offset and sizes to rip audio).

**Comment by Surihix** [comment]
> vgmstream accepts known "real" formats (header + data) and follows info in the header, while psound simply reads raw data (ignoring headers) and does heuristics to autodetect PS-ADPCM data. So, psound playing something isn't a good comparison for us.
> 
> The file you posted isn't correctly ripped, as it starts with the "SShd" header, and ends with "cavia stream format v1.01" + blank data. Proper cavia sound files should start with "cavia stream format v1.01" + blank, and then "SShd". Example: [SNDIMGEU_001.zip](https://github.com/vgmstream/vgmstream/files/9692974/SNDIMGEU_001.zip)
> 
> vgmstream just follows header values, so unless "cavia stream.." goes first data is misaligned (start is set to 0x800, from "cavia stream"). Files also need to be named .ads (SShd's actual extension since .ss2 is fake) but I guess I could allow .ss2 too.
> 
> (there is a table in the .BIN bigfile that gives you proper offset and sizes to rip audio).

Oh I am sorry. should have been a bit specific as to what I did exactly to get the audio track. 

I removed the cavia stream format v1.01 string and the 2008 bytes from that string onwards till the position where the SShd header begins. I thought the data after the cavia stream format string were all null and just copied the data from the SShd string onwards for each track with my extractor program.

I did use that very table from the .BIN file that you mentioned, to extract each track. I did not know much about PSound so thanks for telling me how exactly it works.

Once again extremely sorry for not being specific.

**Comment by Surihix** [comment]
> vgmstream accepts known "real" formats (header + data) and follows info in the header, while psound simply reads raw data (ignoring headers) and does heuristics to autodetect PS-ADPCM data. So, psound playing something isn't a good comparison for us.
> 
> The file you posted isn't correctly ripped, as it starts with the "SShd" header, and ends with "cavia stream format v1.01" + blank data. Proper cavia sound files should start with "cavia stream format v1.01" + blank, and then "SShd". Example: [SNDIMGEU_001.zip](https://github.com/vgmstream/vgmstream/files/9692974/SNDIMGEU_001.zip)
> 
> vgmstream just follows header values, so unless "cavia stream.." goes first data is misaligned (start is set to 0x800, from "cavia stream"). Files also need to be named .ads (SShd's actual extension since .ss2 is fake) but I guess I could allow .ss2 too.
> 
> (there is a table in the .BIN bigfile that gives you proper offset and sizes to rip audio).

I fixed the extraction code on my program by making sure it extracts each audio file with the _cavia stream format v1.01_ string header. vgmstream is now able to play the sound files cleanly.
 
Instead of .ss2 extension, maybe you can add .cads extension support to vgmstream. cads can be abbreviated as cavia audio data stream or cavia's ads.
You can also make vgmstream show some description like Cavia stream format as the cads file is parsed.

**Comment by NicknineTheEagle** [comment]
Are you sure you can't find the format's real extension? If the game's archives don't hold filenames perhaps there's something in the executable. It's best to avoid adding fake extensions when possible.

**Comment by bnnm** [comment]
"cads" is in the bigfile (as BE "sdac"), though not sure if it's really meant to be an extension or just a type identifier. Most of the bigfiles (fpk) have sensical stuff like .lst for text and .pss for videos, but one seems to be a size instead.

**Comment by Surihix** [comment]
> Are you sure you can't find the format's real extension? If the game's archives don't hold filenames perhaps there's something in the executable. It's best to avoid adding fake extensions when possible.

Not sure about the filenames for the audio tracks. its probably stored in another file or hashed in the same bin file itself or does not exist. 


> "cads" is in the bigfile (as BE "sdac"), though not sure if it's really meant to be an extension or just a type identifier. Most of the bigfiles (fpk) have sensical stuff like .lst for text and .pss for videos, but one seems to be a size instead.

Its somewhat a weird but a bit logical format compared to the ones I had seen in other games. 

I do have a rough template that I made for the fpk chunk structure when parsing the data. its experimental but should help in some basic splitting work. 

Some bytes after the fpk header in the start of each bin file or in the sub fpk chunks indicates the number of entries/offsets for each file stored in the bin file or in the sub fpk chunks (relative offsets). 
Then some bytes to indicate the starting position from where the data for all the files begin in the bin file or in the sub fpk chunks, then the offsets for the size of the file from the data's starting position onwards, etc.

Judging by how the extension strings were written for the other files in other bin archive files, the extension is tacked after the size offsets in the entry/offsets section at the starting fpk chunk of the bin file or under the sub fpk chunks. it is written in reverse as sdac due to Big Endian byte ordering and it would be cads. 

Renaming the file to .ads from .cads makes the audio playable by vgmstream and in foobar. 
Here is a track that I extracted with the correct data.
[track2.zip](https://github.com/vgmstream/vgmstream/files/9693663/track2.zip)
